File: src/components/table/columns.ts
/**
 * Pure torrent-table column state.
 *
 * Keeping definitions, order, width clamping, visibility rules, grid derivation
 * and persistence normalization here makes the interaction layer deliberately
 * small and keeps malformed localStorage data away from rendering code.
 *
 * The set is the console's (design frame 1a): a select box, then Name · Size ·
 * Done · Status · Seeds/Peers · Down · Up · ETA · Ratio · Label · Added ·
 * Tracker. Two extras ride along hidden by default — `started` and `finished`,
 * which the desktop shell has always offered and the design has no home for.
 */

export const COLUMN_IDS = [
  "select",
  "name",
  "size",
  "done",
  "status",
  "seedsPeers",
  "down",
  "up",
  "eta",
  "ratio",
  "label",
  "added",
  "tracker",
  "started",
  "finished",
] as const;

export type ColumnId = (typeof COLUMN_IDS)[number];

export interface ColumnDefinition {
  id: ColumnId;
  label: string;
  defaultWidth: number;
  minWidth: number;
  /** Takes the leftover width instead of a fixed one (Name only). */
  flexible?: boolean;
  /** Hidden until the user enables it (optional depth columns). */
  defaultHidden?: boolean;
  /** Right-aligned numeric column. */
  numeric?: boolean;
  /**
   * Fixed to its place: the selection box and the name. Neither can be hidden,
   * resized or moved — a table without a name column is not a table.
   */
  locked?: boolean;
}

/** The design's column set, in its order. */
export const COLUMN_DEFINITIONS: readonly ColumnDefinition[] = [
  { id: "select", label: "", defaultWidth: 26, minWidth: 26, locked: true },
  {
    id: "name",
    label: "Name",
    defaultWidth: 200,
    minWidth: 200,
    flexible: true,
    locked: true,
  },
  { id: "size", label: "Size", defaultWidth: 74, minWidth: 40, numeric: true },
  { id: "done", label: "Done", defaultWidth: 112, minWidth: 64 },
  { id: "status", label: "Status", defaultWidth: 96, minWidth: 64 },
  {
    id: "seedsPeers",
    label: "Seeds/Peers",
    defaultWidth: 66,
    minWidth: 56,
    numeric: true,
  },
  { id: "down", label: "Down", defaultWidth: 78, minWidth: 40, numeric: true },
  { id: "up", label: "Up", defaultWidth: 78, minWidth: 40, numeric: true },
  { id: "eta", label: "ETA", defaultWidth: 64, minWidth: 40, numeric: true },
  {
    id: "ratio",
    label: "Ratio",
    defaultWidth: 54,
    minWidth: 40,
    numeric: true,
  },
  { id: "label", label: "Label", defaultWidth: 88, minWidth: 52 },
  {
    id: "added",
    label: "Added",
    defaultWidth: 86,
    minWidth: 62,
    numeric: true,
  },
  { id: "tracker", label: "Tracker", defaultWidth: 120, minWidth: 70 },
  {
    id: "started",
    label: "Started",
    defaultWidth: 72,
    minWidth: 52,
    defaultHidden: true,
  },
  {
    id: "finished",
    label: "Finished",
    defaultWidth: 72,
    minWidth: 52,
    defaultHidden: true,
  },
];

export interface ColumnState {
  /** Display order; reordering rewrites it. */
  order: ColumnId[];
  widths: Record<ColumnId, number>;
  visibility: Record<ColumnId, boolean>;
}

interface SerializedColumnState {
  version: 1;
  order?: unknown;
  widths: unknown;
  visibility: unknown;
}

const DEFINITIONS_BY_ID = Object.fromEntries(
  COLUMN_DEFINITIONS.map((column) => [column.id, column]),
) as Record<ColumnId, ColumnDefinition>;

export function columnDefinition(id: ColumnId): ColumnDefinition {
  return DEFINITIONS_BY_ID[id];
}

/** A fresh default state (callers may safely mutate their returned copy). */
export function defaultColumnState(): ColumnState {
  return {
    order: [...COLUMN_IDS],
    widths: Object.fromEntries(
      COLUMN_DEFINITIONS.map((column) => [column.id, column.defaultWidth]),
    ) as Record<ColumnId, number>,
    visibility: Object.fromEntries(
      COLUMN_DEFINITIONS.map((column) => [column.id, !column.defaultHidden]),
    ) as Record<ColumnId, boolean>,
  };
}

/** Clamp a width to the column's usable minimum and round to whole pixels. */
export function clampColumnWidth(id: ColumnId, width: number): number {
  const definition = DEFINITIONS_BY_ID[id];
  // A locked column's width is not the user's to change (the selection box).
  if (definition.locked) return definition.defaultWidth;
  const finite = Number.isFinite(width) ? width : definition.defaultWidth;
  return Math.max(definition.minWidth, Math.round(finite));
}

/** Return a new state with one clamped width. */
export function resizeColumn(
  state: ColumnState,
  id: ColumnId,
  width: number,
): ColumnState {
  return {
    ...state,
    widths: { ...state.widths, [id]: clampColumnWidth(id, width) },
  };
}

/** Return a new state with one visibility flag; locked columns stay visible. */
export function setColumnVisible(
  state: ColumnState,
  id: ColumnId,
  visible: boolean,
): ColumnState {
  return {
    ...state,
    visibility: {
      ...state.visibility,
      [id]: DEFINITIONS_BY_ID[id].locked ? true : visible,
    },
  };
}

export function toggleColumn(state: ColumnState, id: ColumnId): ColumnState {
  return setColumnVisible(state, id, !state.visibility[id]);
}

/**
 * Move a column to the slot the dropped-on column occupies.
 *
 * A locked column is never moved, and a column is never moved *past* one: the
 * selection box stays first and the name stays second, so a drop settles at the
 * nearest free slot instead of silently reordering the fixed pair.
 */
export function moveColumn(
  state: ColumnState,
  id: ColumnId,
  before: ColumnId,
): ColumnState {
  if (id === before || DEFINITIONS_BY_ID[id].locked) return state;
  const order = state.order.filter((column) => column !== id);
  const target = order.indexOf(before);
  if (target === -1) return state;

  const lockedBefore = order
    .slice(0, target + 1)
    .filter((column) => DEFINITIONS_BY_ID[column].locked).length;
  order.splice(Math.max(lockedBefore, target), 0, id);
  return { ...state, order };
}

/** Ordered definitions for columns that currently participate in the grid. */
export function visibleColumns(state: ColumnState): ColumnDefinition[] {
  return state.order
    .map((id) => DEFINITIONS_BY_ID[id])
    .filter((definition) => state.visibility[definition.id]);
}

/**
 * CSS grid template. Hidden columns are omitted and Name stays flexible.
 *
 * `only` narrows the grid further for the current viewport (see
 * `useResponsiveColumns`) without changing the saved state.
 */
export function gridTemplateColumns(
  state: ColumnState,
  only?: ColumnId[],
): string {
  return visibleColumns(state)
    .filter((column) => !only || only.includes(column.id))
    .map((column) => {
      const width = clampColumnWidth(column.id, state.widths[column.id]);
      return column.flexible ? `minmax(${width}px, 1fr)` : `${width}px`;
    })
    .join(" ");
}

/** Serialize a normalized, versioned payload for the persisted view object. */
export function serializeColumnState(state: ColumnState): string {
  const payload: SerializedColumnState = {
    version: 1,
    order: state.order,
    widths: state.widths,
    visibility: state.visibility,
  };
  return JSON.stringify(payload);
}

/**
 * Restore persisted state, falling back per field and entirely on bad input.
 * This accepts `unknown` because values originate in localStorage JSON.
 *
 * State saved before the console's column set (a merge of S+P, and Added) still
 * loads: unknown ids are dropped, missing ones appended, and the locked pair
 * pinned to the front.
 */
export function deserializeColumnState(value: unknown): ColumnState {
  if (typeof value !== "string") return defaultColumnState();

  let parsed: unknown;
  try {
    parsed = JSON.parse(value);
  } catch {
    return defaultColumnState();
  }
  if (!isRecord(parsed) || parsed.version !== 1) return defaultColumnState();

  const state = defaultColumnState();
  state.order = normalizeOrder(parsed.order);

  const widths = isRecord(parsed.widths) ? parsed.widths : {};
  const visibility = isRecord(parsed.visibility) ? parsed.visibility : {};
  for (const id of COLUMN_IDS) {
    const width = widths[id];
    if (typeof width === "number")
      state.widths[id] = clampColumnWidth(id, width);

    const visible = visibility[id];
    if (DEFINITIONS_BY_ID[id].locked) state.visibility[id] = true;
    else if (typeof visible === "boolean") state.visibility[id] = visible;
  }

  return state;
}

/**
 * A complete column order: known ids in the given sequence, unknown ones dropped
 * and anything missing appended, so a set that gains or loses a column still
 * loads.
 */
function normalizeOrder(order: unknown): ColumnId[] {
  const requested = Array.isArray(order)
    ? order.filter((id): id is ColumnId =>
        (COLUMN_IDS as readonly string[]).includes(id as string),
      )
    : [];
  const ordered = [...new Set(requested)];
  const missing = COLUMN_IDS.filter((id) => !ordered.includes(id));
  const locked = COLUMN_IDS.filter((id) => DEFINITIONS_BY_ID[id].locked);
  const rest = [...ordered, ...missing].filter((id) => !locked.includes(id));
  return [...locked, ...rest];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: crates/gpui/src/columns.rs
//! Pure torrent-table column state.
//!
//! A port of `src/components/table/columns.ts`. Definitions, width clamping,
//! visibility rules and persistence normalization live here, free of GPUI, so
//! the table's rendering code stays about rendering and malformed persisted
//! data never reaches it.

use crate::table_state::SortColumn;

/// Every column the table knows, in display order.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ColumnId {
    Name,
    Size,
    Done,
    Status,
    Seeds,
    Peers,
    Down,
    Up,
    Eta,
    Ratio,
    Label,
    Tracker,
    Started,
    Finished,
}

impl ColumnId {
    /// Display order, which is also the persisted order.
    pub const ALL: [Self; 14] = [
        Self::Name,
        Self::Size,
        Self::Done,
        Self::Status,
        Self::Seeds,
        Self::Peers,
        Self::Down,
        Self::Up,
        Self::Eta,
        Self::Ratio,
        Self::Label,
        Self::Tracker,
        Self::Started,
        Self::Finished,
    ];

    #[must_use]
    pub const fn index(self) -> usize {
        self as usize
    }

    /// The stable key used in `settings.json`.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Name => "name",
            Self::Size => "size",
            Self::Done => "done",
            Self::Status => "status",
            Self::Seeds => "seeds",
            Self::Peers => "peers",
            Self::Down => "down",
            Self::Up => "up",
            Self::Eta => "eta",
            Self::Ratio => "ratio",
            Self::Label => "label",
            Self::Tracker => "tracker",
            Self::Started => "started",
            Self::Finished => "finished",
        }
    }

    #[must_use]
    pub fn parse(value: &str) -> Option<Self> {
        Self::ALL.into_iter().find(|id| id.as_str() == value)
    }
}

/// One column's fixed properties.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct ColumnDefinition {
    pub id: ColumnId,
    pub label: &'static str,
    pub default_width: f32,
    pub min_width: f32,
    /// Takes the leftover width instead of a fixed width (Name only).
    pub flexible: bool,
    /// Hidden until the user enables it, rather than shown on first run.
    pub default_hidden: bool,
    /// Site of the header click that sorts by this column, when it sorts.
    pub sort: Option<SortColumn>,
}

/// The definition table, matching `COLUMN_DEFINITIONS`.
pub const COLUMN_DEFINITIONS: [ColumnDefinition; 14] = [
    ColumnDefinition {
        id: ColumnId::Name,
        label: "Name",
        default_width: 200.,
        min_width: 120.,
        flexible: true,
        default_hidden: false,
        sort: Some(SortColumn::Name),
    },
    ColumnDefinition {
        id: ColumnId::Size,
        label: "Size",
        default_width: 70.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: Some(SortColumn::Size),
    },
    ColumnDefinition {
        id: ColumnId::Done,
        label: "Done",
        default_width: 92.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: Some(SortColumn::Percent),
    },
    ColumnDefinition {
        id: ColumnId::Status,
        label: "Status",
        default_width: 84.,
        min_width: 64.,
        flexible: false,
        default_hidden: false,
        sort: Some(SortColumn::Status),
    },
    ColumnDefinition {
        id: ColumnId::Seeds,
        label: "S",
        default_width: 52.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: None,
    },
    ColumnDefinition {
        id: ColumnId::Peers,
        label: "P",
        default_width: 52.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: None,
    },
    ColumnDefinition {
        id: ColumnId::Down,
        label: "Down",
        default_width: 76.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: Some(SortColumn::DownRate),
    },
    ColumnDefinition {
        id: ColumnId::Up,
        label: "Up",
        default_width: 76.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: Some(SortColumn::UpRate),
    },
    ColumnDefinition {
        id: ColumnId::Eta,
        label: "ETA",
        default_width: 62.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: Some(SortColumn::EtaSeconds),
    },
    ColumnDefinition {
        id: ColumnId::Ratio,
        label: "Ratio",
        default_width: 46.,
        min_width: 40.,
        flexible: false,
        default_hidden: false,
        sort: Some(SortColumn::Ratio),
    },
    ColumnDefinition {
        id: ColumnId::Label,
        label: "Label",
        default_width: 72.,
        min_width: 52.,
        flexible: false,
        default_hidden: false,
        sort: None,
    },
    ColumnDefinition {
        id: ColumnId::Tracker,
        label: "Tracker",
        default_width: 110.,
        min_width: 70.,
        flexible: false,
        default_hidden: false,
        sort: None,
    },
    ColumnDefinition {
        id: ColumnId::Started,
        label: "Started",
        default_width: 72.,
        min_width: 52.,
        flexible: false,
        default_hidden: true,
        sort: Some(SortColumn::StartedAt),
    },
    ColumnDefinition {
        id: ColumnId::Finished,
        label: "Finished",
        default_width: 72.,
        min_width: 52.,
        flexible: false,
        default_hidden: true,
        sort: Some(SortColumn::FinishedAt),
    },
];

/// The definition of one column.
#[must_use]
pub fn definition(id: ColumnId) -> ColumnDefinition {
    COLUMN_DEFINITIONS[id.index()]
}

/// One column's persisted width and visibility.
#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ColumnPref {
    pub id: String,
    pub width: f32,
    pub visible: bool,
}

/// The table's live column state.
#[derive(Clone, Debug, PartialEq)]
pub struct ColumnState {
    widths: [f32; 14],
    visibility: [bool; 14],
}

impl Default for ColumnState {
    fn default() -> Self {
        let mut widths = [0.; 14];
        let mut visibility = [false; 14];
        for id in ColumnId::ALL {
            let definition = definition(id);
            widths[id.index()] = definition.default_width;
            visibility[id.index()] = !definition.default_hidden;
        }
        Self { widths, visibility }
    }
}

impl ColumnState {
    /// Clamp a width to its column's usable minimum, rounding to whole pixels.
    #[must_use]
    pub fn clamp_width(id: ColumnId, width: f32) -> f32 {
        let definition = definition(id);
        if width.is_finite() {
            width.max(definition.min_width).round()
        } else {
            definition.default_width
        }
    }

    #[must_use]
    pub fn width(&self, id: ColumnId) -> f32 {
        Self::clamp_width(id, self.widths[id.index()])
    }

    #[must_use]
    pub fn is_visible(&self, id: ColumnId) -> bool {
        // Name is the row's identity and cannot be hidden.
        id == ColumnId::Name || self.visibility[id.index()]
    }

    /// Set one column's width, clamped.
    pub fn resize(&mut self, id: ColumnId, width: f32) {
        self.widths[id.index()] = Self::clamp_width(id, width);
    }

    /// Set one column's visibility; Name stays visible.
    pub fn set_visible(&mut self, id: ColumnId, visible: bool) {
        if id == ColumnId::Name {
            self.visibility[id.index()] = true;
            return;
        }
        self.visibility[id.index()] = visible;
    }

    pub fn toggle(&mut self, id: ColumnId) {
        let next = !self.is_visible(id);
        self.set_visible(id, next);
    }

    /// Restore every column to its definition defaults.
    pub fn reset(&mut self) {
        *self = Self::default();
    }

    /// The columns currently in the grid, in display order.
    pub fn visible(&self) -> Vec<ColumnId> {
        ColumnId::ALL
            .into_iter()
            .filter(|id| self.is_visible(*id))
            .collect()
    }

    /// The persisted form: one entry per column, always in display order so a
    /// settings diff reads sensibly.
    #[must_use]
    pub fn prefs(&self) -> Vec<ColumnPref> {
        ColumnId::ALL
            .into_iter()
            .map(|id| ColumnPref {
                id: id.as_str().to_owned(),
                width: self.width(id),
                visible: self.is_visible(id),
            })
            .collect()
    }

    /// Rebuild from persisted state, falling back per field and entirely on
    /// unknown input, as `deserializeColumnState` does in the TypeScript tree.
    #[must_use]
    pub fn from_prefs(prefs: &[ColumnPref]) -> Self {
        let mut state = Self::default();
        for pref in prefs {
            let Some(id) = ColumnId::parse(&pref.id) else {
                continue;
            };
            state.resize(id, pref.width);
            state.set_visible(id, pref.visible);
        }
        state
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_hide_only_the_date_columns() {
        let state = ColumnState::default();
        assert!(!state.is_visible(ColumnId::Started));
        assert!(!state.is_visible(ColumnId::Finished));
        assert_eq!(state.visible().len(), 12);
        assert_eq!(state.visible()[0], ColumnId::Name);
    }

    #[test]
    fn name_cannot_be_hidden() {
        let mut state = ColumnState::default();
        state.set_visible(ColumnId::Name, false);
        state.toggle(ColumnId::Name);
        assert!(state.is_visible(ColumnId::Name));
    }

    #[test]
    fn widths_clamp_to_the_definition() {
        let mut state = ColumnState::default();
        state.resize(ColumnId::Name, 10.);
        assert_eq!(state.width(ColumnId::Name), 120.);
        state.resize(ColumnId::Size, 123.4);
        assert_eq!(state.width(ColumnId::Size), 123.);
        state.resize(ColumnId::Size, f32::NAN);
        assert_eq!(state.width(ColumnId::Size), 70.);
    }

    #[test]
    fn prefs_round_trip_and_ignore_unknown_columns() {
        let mut state = ColumnState::default();
        state.resize(ColumnId::Tracker, 200.);
        state.set_visible(ColumnId::Started, true);
        state.set_visible(ColumnId::Ratio, false);

        let mut prefs = state.prefs();
        prefs.push(ColumnPref {
            id: "nonexistent".into(),
            width: 40.,
            visible: true,
        });
        let restored = ColumnState::from_prefs(&prefs);
        assert_eq!(restored, state);
        assert!(restored.is_visible(ColumnId::Started));
        assert!(!restored.is_visible(ColumnId::Ratio));
        assert_eq!(restored.visible().len(), 12);
    }

    #[test]
    fn only_numeric_columns_are_sortable_headers() {
        assert_eq!(definition(ColumnId::Done).sort, Some(SortColumn::Percent));
        assert_eq!(
            definition(ColumnId::Finished).sort,
            Some(SortColumn::FinishedAt)
        );
        assert_eq!(definition(ColumnId::Seeds).sort, None);
        assert_eq!(definition(ColumnId::Tracker).sort, None);
    }
}
